### ISSUE TITLE
Terminal Providers

### ISSUE THREAD

**Issue by IlyaBiryukov** [Microsoft/vscode#46192]
Issue Type: <b>Feature Request</b>

Enable extensions provide custom terminals to other extensions and tasks

Today extensions and tasks use vscode's API `vscode.window.createTerminal(terminalOptions)` to create a new terminal window, and some other APIs to render task output as terminal output.
Our extension can change the way a terminal is started, and we want user to be able to apply this change to other terminals and tasks vscode may be running without changing their code.

The idea is to introduce a concept of _Terminal Provider_, an object that will be chained when something starts a new terminal:

```typescript
interface TerminalProvider {
    createTerminal(options: TerminalOptions, provider: TerminalProvider): Promise<Terminal>;
}

// A way to register a terminal provider. This function returns a disposable
// which, when disposed, unregisters the provider.
window.registerTerminalProvider(provider: TerminalProvider): Disposable;
```

When a provider is registered, it will be called when anyone calls `window.createTerminal(...)`. The `provider` argument in `createTerminal` is the previous chained provider, or internal vscode implementation that actually creates a terminal.

Registering or un-registering a provider should not affect existing terminals.

**Comment by dbaeumer** [comment]
As @Tyriar pointed out in mail this is something we discussed this week as well, especially in the use case of tasks. I would like to suggest a different form of API instead of a terminal provider mainly because it will be unclear if two extensions register such a terminal provider. (side note: I looked into a task execution provider for a comparable use case this milestone and couldn't find a nice working solution there either).

I would like to see three different sets of API around terminals that extensions then can use to control how a terminal is functioning:

- events about the existence of a terminal process (e.g. the node-pty process in VS Code). The events would provide access to stdout and stdin of that process either via streams or Event<string>.
- API to create a terminal renderer without spawning a underlying terminal process (eg. node-pty processs). The returned renderer would offer methods to write and read from the terminal and events when it gets resized. Not sure if we need special handling for Ctrl+C or other Control characters.
- API to create a terminal process (e.g. the node-pty process) and access to its stdin and stdout stream. Here I am not sure if this is really necessary or if we ask developers to simply use child_process functions. Providing API ensure that for node-pty we can reuse our tested version of node-pty.

Combining these three API would allow to implement a `TerminalProvider` concept.

**Comment by Tyriar** [comment]
> Not sure if we need special handling for Ctrl+C or other Control characters.

These would be translated by automatically and passed through the read event in this case.

> API to create a terminal process

`child_process` is insufficient as it doesn't have dimensions.

**Comment by IlyaBiryukov** [comment]
@dbaeumer The main reason for this feature request for us is to be able to "hijack" or "substitute" a terminal that other extensions would start without changing their code. They use `vscode.window.createTerminal`. The terminal provider registration should be able to affect how `vscode.window.createTerminal` works.

**Comment by dbaeumer** [comment]
@IlyaBiryukov but your provider is a one of solution. If a second extensions wants to do this it either can't or replaces your hijack terminal in a way that your provider is lost. The proposed API would allow for equal participation on the output / input stream. This allows for `hijacking` the input and output only but to my understanding this should be ok in the use cases I am aware of.

**Comment by IlyaBiryukov** [comment]
@dbaeumer My idea was to form a chain of providers, when they fire one-by-one, and and each pass `vscode.TerminalOptions` to the next in line. So registering a new provider won't replace or hide previous ones, it'll just add another step in the processing. 
That being said, I'm not opposed to having more granular control on terminals and the ability to hijack input and output streams.

**Comment by dbaeumer** [comment]
@IlyaBiryukov make sense.

**Comment by Tyriar** [comment]
@IlyaBiryukov this is what we've come up with so far, I'm interested in your thoughts:

```ts
export namespace window {
    // A Terminal owns a process and a renderer
    export interface Terminal {
        // This is unchanged
        readonly name: string;

        // processId not needs to also return undefined when there is no backing
        // pid (TerminalRenderer)
        readonly processId: Thenable<number | undefined>;

        // Send text to process
        sendText(data: string): void; // in

        // Data from the process
        onData(cb: (data: string) => void): void; // out

        // Overrides the terminal renderer size, the size can be less than or
        // equal to the actual size allowed by the panel
        setSize(cols: number, rows: number): void;

        // Fires when the panel area is resized, this fires even when setSize
        // has been called
        onResize(cb: (cols: number, rows: number) => void): void;
    }

    // Creates and returns an empty TerminalRenderer
    export function createTerminalRenderer(): TerminalRenderer;
    
    // A TerminalRender does not own a process, it's similar to an output window
    // but it understands ANSI
    export interface TerminalRenderer {
        // Extensions can set the name (what appears in the dropdown)
        name: string;

        // Write to xterm.js
        write(data: string): void; // out

        // key press, or sendText was triggered by an extension on the terminal
        onData(cb: (data: string) => void): void; // in

        // Overrides the terminal renderer size, the size can be less than or
        // equal to the actual size allowed by the panel
        setSize(cols: number, rows: number): void;

        // Fires when the panel area is resized, this fires even when setSize
        // has been called
        onResize(cb: (cols: number, rows: number) => void): void;
    }

    // Allows listening to terminal creation events, when a TerminalRenderer is
    // created in this manner is uses Terminal as a facade
    export function onDidCreateTerminal: Event<Terminal>;

    // Do not allow interception/cancellation of terminals, we do not want to
    // encourage extensions to do this. Extensions also treat TerminalOptions
    // as very environment-specific
}
```

An example extension: When a terminal is created it creates a another terminal which mirrors it and accepts input:

```ts
let wasMirrored = false;
const linkedTerminals: {[terminal: Terminal]: TerminalRenderer} = {};

onDidCreateTerminal(terminal => {
    if (wasMirrored) {
        // Ignore events that came from mirrored terminals
        wasMirrored = false;
    } else {
        wasMirrored = true;
        linkedTerminals[terminal] = window.createTerminalRenderer();
        setupListeners(terminal, linkedTerminals[terminal]);
    }
});

function setupListeners(master: Terminal, slave: TerminalRenderer) {
    // Send print instructions to slave
    master.onData(data => slave.write(data));
    //  Redirect slave input to master
    slave.onData(data => master.sendText(data));
}
```

**Comment by Tyriar** [comment]
I updated inline with feedback from @jrieken:

- `onResize`, `onData` is an event
- Size functions now work on `TerminalDimensions`
- `name` is a mandatory `createTerminalRenderer` parameter

**Comment by Tyriar** [comment]
Added `terminals: Terminal[]` for completeness/consistency with rest of API (from https://github.com/Microsoft/vscode/issues/13267).

**Comment by IlyaBiryukov** [comment]
@Tyriar, here is my feedback on the proposed API:
1. How one can figure out the actual size allowed by the panel (the maximum for the terminal dimensions allowed by `setSize`)? Can this maximum sizes change over time, e.g. when user resizes the window?
2. If something wants to create a terminal, the way to intercept it is to listen to `onDidCreateTerminal` and then create terminal renderer? How will VSCode know that it doesn't need to create its own renderer?
3. About this:
   `// Do not allow interception/cancellation of terminal`
   Isn't `onDidCreateTerminal` a way to intercept creation of a terminal?
4. About `createTerminalRenderer`, what about specifying the dimensions, and whether the renderer has fixed dimensions or fit the window?
5. About `TerminalRenderer` interface, how one can close it? Say that underlying terminal has closed, how the renderer would know that and close itself?
6. The example extension, will it only work for the first terminal, and then will stop working because `wasMirrored` will become `true`?
7. Are `Terminal` and `TerminalRenderer` persistent objects, can `===` be used on them?
8. Why `processId` is needed on `Terminal`? 
9. For `Terminal` object, is there a way to query its current dimensions?
10. Can 'onData' event handler block the terminal from producing more data? The scenario is that the consumers can be slow to render the data, and the terminal may be producing lots of data fast. Can `onData` block the terminal from producing more data? Considering that there may be more than one handler there (e.g. a natural handler, and the one that broadcasts the data to other participants).
11. For 'TerminalRenderer', and 'Terminal' we will need `onDidClose` event to know when they're gone.

**Comment by Tyriar** [comment]
@IlyaBiryukov 

1. `onResize` will fire any time the window/panel is resized such that the terminal cols/rows change. So you can do this by tracking these `onResize` events.
2. > How will VSCode know that it doesn't need to create its own renderer?

   Don't totally understand this, a single terminal in the dropdown can either be backed by a single `Terminal` (which has a process) or a single `TerminalRenderer` (which does not), not both.
3. The comment is a little ambiguous, you can listen to terminal creation but you cannot stop it as the team doesn't want to encourage extensions to do this.
4. You could do this with:

   ```ts
   const term = createTerminalRenderer('me');
   term.setSize(x, y);
   ```
5. Good question, it should probably also have `dispose`.
6. `wasMirrored` was there so the creation events would be ignored and stop a stack overflow.
7. No you cannot compare them and there is no reliable way to do this.
8. `Terminal.processId` already exists, the change is to add `undefined` when the `Terminal` is backed only by a `TerminalRenderer`.
9. No, see 1
10. I think this is fine from your perspective, we can go into more detail in email.
11. This already exists, but I'm beginning to feel we need some way of comparing `Terminal` and `TerminalRenderer` though.

**Comment by DanTup** [comment]
Could this help with #46696? Would it allow us to manipulate the `PATH` being set for a terminal from an extension?

**Comment by dbaeumer** [comment]
@weinand FYI.

**Comment by Tyriar** [comment]
@DanTup no you cannot manipulate a terminal's environment that was launched (`onDidCreateTerminal` happened after it was already created).

**Comment by DanTup** [comment]
Understood. Though maybe this pattern would work there - if this was extended to have an onWillCreateTerminal that could modify the environment, it might fix that issue? (I don't want to derail this though, so if it sounds reasonable I could add comments to the other issue suggesting it?)

**Comment by IlyaBiryukov** [comment]
@Tyriar Thanks for reviewing my feedback. More below:
1. It looks doable, but rather awkward to watch for `onResize` event just to figure out the current size.
2. It is still not clear how to figure out max dimensions that we can use for `TerminalRenderer.setSize()`. 
   Why there is this max dimensions limit?
3. So, after calling `TerminalRenderer.setSize()` it becomes fixed-sized and doesn't auto-fit the window it is rendered in when user resizes VSCode? Or does it change the whole VSCode UI layout, like resize the whole terminal pane? 

Here I'm assuming `TerminalRenderer` is the front end for the xterm control, and the terminal pane is where VSCode renders it, the tool window with the drop down for other terminals, and the buttons to control the terminals.

**Comment by Tyriar** [comment]
@IlyaBiryukov 

1. It's just keeping track of 2 numbers isn't it? Perhaps we should expose `Terminal`/`TerminalRenderer.dimensions: TerminalDimensions`?
2. The max dimensions limit is a way for us to prevent extensions adding scroll bars while also fitting into your use case.
3. The terminal will no longer auto fit to the panel, the panel/window itself won't change, just the number of columns and rows within the terminal.

**Comment by alpaix** [comment]
@Tyriar @dbaeumer 
Would it be possible to identify terminals used for rendering of task output?
And how could an extension associate a terminal output with a `TaskExecution` object?

**Comment by IlyaBiryukov** [comment]
@Tyriar 
1. Yep, exposing `dimensions` will help.
2. So max dimensions are determined by the size of the terminal pane, and will change when user resizes VSCode window? How one can get these max dimensions? I keep asking that :) E.g. in Console in .NET has [LargestWindowWidth property](https://msdn.microsoft.com/en-us/library/system.console.largestwindowwidth(v=vs.110).aspx)
3. Will the panel/window have fixed size when user resizes VSCode window in this case, after an extension called `setSize()` on `Terminal` or `TerminalRenderer`? Let my clarify this a bit. Is my understanding correct in the following use case:
    1. An extension calls `Terminal.setSize()` or `TerminalRenderer.setSize()` with, let's say dimensions of 100 columns and 50 rows. Let's assume its a valid dimension, less than the max allowed.
    2. The terminal and terminal renderer control now has 100 columns and 50 rows
    3. User resizes VSCode window, let's say reduces its size
    4. Will the terminal or terminal renderer dimension stay the same, 100 x 50?
    5. Will the scroll bars appear if the terminal renderer size doesn't fit the new window size?
   

cc @alpaix

**Comment by dbaeumer** [comment]
@alpaix why do you need this information? I am a little reluctant to strictly associate tasks with terminals in the API.

**Comment by dbaeumer** [comment]
I am still not understanding why you would need to. In my understanding the task API is used to control task (query, execute, terminate them). When a tasks gets executed you can use the terminal API to remote the output. 

The only way how we today associated them in the user interface is by having 'Task' in the terminal name:

![capture](https://user-images.githubusercontent.com/1931590/38665432-aa8c355c-3e3c-11e8-8e2b-82838ffbea89.png)

**Comment by alpaix** [comment]
This approach only works partially in the vscode-to-vscode remoting scenario. In this design, terminals will be remoted in a task agnostic fashion. Then the remote vscode will be provided with a set of tasks mirroring the host ones. When a task is executed vscode will create a new terminal for the mirror task as it would do for any other task. As the result of this operation, the user will get two terminals per task in the remote client, which is not the best UX.

Also, this approach will not work in a hybrid setup, when the remote client is a different IDE, e.g. VisualStudio, and the task output needs to be rendered not as a terminal to provide "native" UX.

**Comment by dbaeumer** [comment]
I see the problems for the VS guest UI. However if the guest is VS Code I would simply wait until the terminal is mirrored. Why does the guest VS need to create the terminal in this case?

**Comment by dbaeumer** [comment]
In general I think the guest shouldn't make the assumption that a task is executed in the terminal since this is something the host decides (what if VS Code moves away from executing tasks in the terminal ?). Should the host return something back to the guest that lets the guest know how to best render the output. In the VS Code <-> VS Code case the result would be: do nothing. In the VS <-> VS Code case we need to find a way to render the task output provided by a terminal in Visual Studio. I will think how this can best be done without marrying tasks with terminals on the VS Code side.

**Comment by alpaix** [comment]
I agree this is not a trivial problem. Initially, I planned to "hack" somehow the task execution mechanism on the host to inject a shim process which would tee the task process output for remoting.

**Comment by dbaeumer** [comment]
One first idea is to include the process ID in the task execution. Tasks as they are defined in VS Code are external processes. So this is something I think can be kept stable.

**Comment by Tyriar** [comment]
> Also, this approach will not work in a hybrid setup, when the remote client is a different IDE, e.g. VisualStudio, and the task output needs to be rendered not as a terminal to provide "native" UX.

Could you special case the VS host scenario to basically mirror the task output in a `TerminalRenderer`, leveraging the terminal sharing code? If tasks move away from the terminal that's a problem we can tackle then?

**Comment by IlyaBiryukov** [comment]
@Tyriar thanks for elaborating on dimensions
* Is the terminal/terminal renderer initially created to fit the window and its dimensions will match the window size?
* Will there be `onDidChangeDimensions` fired for the terminal just after the creation that will contain the current dimensions that are also the maximum dimensions?
* Will there be an `Error` thrown if somebody tried to set `dimensions` to something more than their current maximum?
* When user resizes VSCode window, and it gets larger, will it keep the dimensions if there were set before and still fit the new window size?
* When user resizes VSCode window, and one of the dimensions is still ok, but the other one needs to be adjusted, what will happen? Will the dimensions snap to the window size then?
* Given that the terminal already has vertical scroll bar, does it mean that there is a less strict restriction on the `rows` dimension?

Also cc @dgriffen for whack-whack terminal APIs.

**Comment by Tyriar** [comment]
@IlyaBiryukov 

> Is the terminal/terminal renderer initially created to fit the window and its dimensions will match the window size?

Yes, the behavior is the same as today unless you set `dimensions`.

> Will there be onDidChangeDimensions fired for the terminal just after the creation that will contain the current dimensions that are also the maximum dimensions?

No, you will probably need to do this:

```ts
const term = createTerminal();
// Make sure the process is ready
term.processId(() => {
  term.dimensions
});
```

> Will there be an Error thrown if somebody tried to set dimensions to something more than their current maximum?

I will need to check how bad args are handled elsewhere, it should act consistently (either throwing an error or you need to check dimensions again?).

> When user resizes VSCode window, and it gets larger, will it keep the dimensions if there were set before and still fit the new window size?

Yes, when `dimensions` is set, only when they do not fit will it be forced to become smaller.

> When user resizes VSCode window, and one of the dimensions is still ok, but the other one needs to be adjusted, what will happen? Will the dimensions snap to the window size then?

Good edge case, yes I think they would just reset to the maximum in that case.

> Given that the terminal already has vertical scroll bar, does it mean that there is a less strict restriction on the rows dimension?

No because the scroll bar scroll the _scrollback_, not the _viewport_. If larger rows were allowed that would need 2 vertical scroll bars side-by-side 🤢.

**Comment by dgriffen** [comment]
@IlyaBiryukov currently the \\terminal APIs don't have any way to report or control the size. The size is set to the maximum possible upon creation and adjusted as needed. Size adjustments are send into node-pty as a backend so I imagine there is some way for the launched terminal process to access that information.

**Comment by DonJayamanne** [comment]
I'm considering using this new API in our Python extension.
Currently we have a custom command in the python extension that creates a terminal (so we can execute some scripts, required to activate a python environment, much like `nvm use`.

With this new API we can now intercept creation of terminals and then send the custom text. So any terminal created will always be activated. I.e. users will not have to use any special (custom) commands. They could use the (+) icon too.

However, there's a problem I can see with this approach. Well end up intercepting the terminals created by our own extension (hence we'd end up sending the commands twice), similarly we could end up sending commands to terminals created by other extensions!!

Is there a way to identify terminals created by the user vs created via extension APIs.

**Comment by Tyriar** [comment]
@DonJayamanne 

> With this new API we can now intercept creation of terminals and then send the custom text. So any terminal created will always be activated. I.e. users will not have to use any special (custom) commands. They could use the (+) icon too.

Do you mean to amend the environment? You will run into problems of not knowing what to run to set the environment. it may not be clear from the executable. It could also be something like a task, a debug console or some other special extension use case which might break those.

> Is there a way to identify terminals created by the user vs created via extension APIs.

```ts
const myTerm = createTerminal();
window.onDidOpenTerminal(t => {
  if (t !== myTerm) {
    // do stuff here
  }
});
```

**Comment by alpaix** [comment]
> Could you special case the VS host scenario to basically mirror the task output in a `TerminalRenderer`, leveraging the terminal sharing code?

By adopting this approach, it's going to be a core principle for both IDE client implementations. It will require knowing how to distinguish a task terminal from an interactive one; mainly because we'll need to apply different sharing permissions and policies. For instance, a task terminal will be remoted automatically in a read-only way, while sharing an interactive terminal will require a host user to issue a command to share one.

>  Tasks as they are defined in VS Code are external processes.

I suppose process ID of the terminal is not the same as the process ID of the task. And if I understand correctly, tasks will be rendered in a built-in in-proc terminal, correct?

> One first idea is to include the process ID in the task execution. 

If the task process ID won't help me to associate it with a terminal linked to the task execution, then having this process ID attribute on the task execution will be redundant.

To summarize, with this design it's utterly necessary to determine a category of a terminal because these categories are conceptually different from a user's perspective and therefore need to be handled separately.

**Comment by alpaix** [comment]
@Tyriar 
Relying solely on the terminal name is probably not the best idea. First, because the naming pattern is not part of the supported API it could be easily broken when you guys decide to change the naming convention. Second, it's not clear how to compare the name in the case of a localized application instance. And third, a 3rd party extension could create a terminal with a similar name format which would cause even more confusion.

**Comment by Tyriar** [comment]
@alpaix My example wouldn't break, it would just introduce a minor inconsistency with the naming of the task. That is about matching a renderer which you make to the corresponding terminal, for the case where the task is launching from VS. That's the case we're talking about isn't it?

The listening for the name is also just double checking, `onDidOpenTerminal` will fire within milliseconds of `createTerminalRenderer` being called.

**Comment by Tyriar** [comment]
@IlyaBiryukov @alpaix I ust pushed the `Terminal.onData` API, see example: https://github.com/Microsoft/vscode-extension-samples/commit/b3babe0eeed1dd4ca2d4cde8b14c4f16441aeb43

**Comment by alpaix** [comment]
@Tyriar Thanks for pushing this change. I tested it by running the `terminal-sample` extension with the latest VSCode insiders. It works well for interactive terminals. Unfortunately, it seems to fail to capture output from a task terminal. Is there any limitation that prevents certain v2.0 tasks to be handled correctly.

**Comment by Tyriar** [comment]
@alpaix the extension command works for me on VS Code's build task:

<img width="670" alt="screen shot 2018-04-24 at 4 57 45 pm" src="https://user-images.githubusercontent.com/2193314/39219633-a6f47776-47e0-11e8-9769-861f9725333a.png">

**Comment by alpaix** [comment]
I tried running it on my Windows desktop. I have a project with the following task:
```js
{
    "version": "2.0.0",
    "tasks": [
        {
            "label": "build",
            "type": "shell",
            "command": "npm",
            "args": [
                "run",
                "compile"
            ],
            "problemMatcher": "$tsc-watch",
            "group": "build",
            "presentation": {
                "echo": true,
                "reveal": "always",
                "focus": false,
                "panel": "shared"
            }
        }
    ]
}
```
Then I invoked _Terminal API: Attach data listener_ command and executed a task. No data have been passed to the listener.

At the same time, it works perfectly well when I attached a listener to a regular Powershell terminal.

**Comment by Tyriar** [comment]
@alpaix you need to use `onData` on the specific `Terminal`, which can only be done after you have started the task.

**Comment by alpaix** [comment]
The _Attach_ command asks for a concrete terminal instance to get attached to. So it does find the task terminal instance and adds a listener to the right terminal.

I tweaked the sample extension to attach the listener to every new terminal:
```ts
// vscode.window.onDidOpenTerminal
	if ('onDidOpenTerminal' in vscode.window) {
		(<any>vscode.window).onDidOpenTerminal((terminal: vscode.Terminal) => {
			vscode.window.showInformationMessage(`onDidOpenTerminal, name: ${terminal.name}`);
			(<any>terminal).onData((data: string) => {
				console.log('onData: ' + data);
			});
		});
	}
```

After this change it started getting the data. Is the right timing critical? Is it possible a listener attached later won't get any data?

**Comment by Tyriar** [comment]
@alpaix it should work later as well. When onData is called the extension host tells the renderer process to start sending data events:

https://github.com/Microsoft/vscode/blob/e5f33576004c4c2bf07af3a5661d47df446468c0/src/vs/workbench/api/electron-browser/mainThreadTerminalService.ts#L91-L97

It should fire the data event until the terminal instance is disposed/exited.

**Comment by dbaeumer** [comment]
I think @alpaix is asking what happpens to data that is already present in the terminal. Will it be sent as well or will it be lost in terms of events.

**Comment by Tyriar** [comment]
> Does the implementation expect listeners to get attached within a short period of time after the terminal instance is created? Based on my observation, a listener attached later simply won't get any data,

@alpaix yes this is the case.

**Comment by Tyriar** [comment]
@IlyaBiryukov so this is done, check the proposed API file for the latest:

https://github.com/Microsoft/vscode/blob/442e5e202a89c1f8efe642c89c0554cab1f15a9e/src/vs/vscode.proposed.d.ts#L362-L383

I made a few tweaks on the initial design, namely:

- `TerminalRenderer.dimensions` is only the "overridden" dimensions
- `TerminalRenderer.maximumDimensions` always represents the maximum dimensions
- `TerminalRenderer.onDidChangeMaximumDimensions` fires whenever `maximumDimensions` is updated
- `TerminalRenderer.terminal` is a convenience property to fetch the underlying `Terminal`, I felt this was needed as it was super inconvenient to call `Terminal.show` for example and didn't want to duplicate things across the 2 interfaces. Note that this is a `Thenable<Terminal>`
- `TerminalRenderer.onData` was renamed to `onInput` to make it more obvious that it's different to `Terminal.onData`
- Dimensions are not exposed on `Terminal` as I only want the creator of the `TerminalRenderer` to change this

Any feedback on this would be great 😃. I've added pretty extensive doc comments so they're worth a read before trying it out.

**Comment by IlyaBiryukov** [comment]
@Tyriar Excellent news! I like the new API very much and look forward to playing with it. Will it be shipping in the next insiders update?

**Comment by Tyriar** [comment]
@IlyaBiryukov I think it's already in the latest, if not it will be in tomorrow's.

**Comment by Tyriar** [comment]
@IlyaBiryukov also the samples I was working against while developing are in this commit: https://github.com/Microsoft/vscode-extension-samples/commit/5f23d63fd152ae4900f2c399b64f3d0d3fb097a5

**Comment by IlyaBiryukov** [comment]
@Tyriar is the `TerminalRenderer` API stable in insiders build *1.25.0*, or there will be more changes?
I adopted the terminal renderer APIs from that build and everything works like a charm.
